src/MigrationPersistor.js

```
@@ -0,0 +1,232 @@
+          if (Settings.fallback.copyOnMiss) {
+            const fallbackStream = await fallback.promises.getObjectStream(
+              fallbackBucket,
+              key,
+              {}
+            )
+            // run in background
+            _copyStreamFromFallbackAndVerify(
+              fallbackStream,
+              fallbackBucket,
+              bucket,
+              key,
+              key
+            ).catch((err) => {
+              Logger.warn({ err }, 'failed to copy file from fallback')
+            })
+          }
+          return fallback.promises[method](fallbackBucket, key, ...moreArgs)
+        }
+        throw err
+      }
+    }
+  }
+
+  async function _copyStreamFromFallbackAndVerify(
+    stream,
+    sourceBucket,
+    destBucket,
+    sourceKey,
+    destKey
+  ) {
+    try {
+      let sourceMd5
+      try {
+        sourceMd5 = await fallback.promises.getObjectMd5Hash(
+          sourceBucket,
+          sourceKey
+        )
+      } catch (err) {
+        Logger.warn(err, 'error getting md5 hash from fallback persistor')
+      }
+
+      await primary.promises.sendStream(destBucket, destKey, stream, sourceMd5)
+    } catch (err) {
+      const error = new WriteError({
+        message: 'unable to copy file to destination persistor',
+        info: {
+          sourceBucket,
+          destBucket,
+          sourceKey,
+          destKey
+        }
+      }).withCause(err)
+      if (Settings.Metrics) {
+        Settings.Metrics.inc('fallback.copy.failure')
+      }
+
+      try {
+        await primary.promises.deleteObject(destBucket, destKey)
+      } catch (err) {
+        error.info.cleanupError = new WriteError({
+          message: 'unable to clean up destination copy artifact',
+          info: {
+            destBucket,
+            destKey
+          }
+        }).withCause(err)
+      }
+
+      Logger.warn({ error }, 'failed to copy file from fallback')
+      throw error
```

Rather than logging and throwing, we could copy this logging statement in all the empty catch handlers above. It would prevent logging twice in the case where this error is caught at a higher level.